tools/gomod-local-update/internal/updater/module_operator.go

```diff
@@ -0,0 +1,154 @@
+// Returns "v2" for /v2, "v1" for no version suffix
+func getMajorVersion(modulePath string) string {
+	re := regexp.MustCompile(majorVersionPattern)
+	if match := re.FindString(modulePath); match != "" {
+		return strings.TrimPrefix(match, "/")
+	}
+	return "v0"
+}
+
+type moduleOperator struct {
+	config *Config
+}
+
+// NewModuleOperator creates a new ModuleOperator
+func NewModuleOperator(config *Config) ModuleOperator {
+	if config.RepoRemote == "" {
+		config.RepoRemote = "origin"
+	}
+	if config.BranchTrunk == "" {
+		config.BranchTrunk = "develop"
+	}
+	return &moduleOperator{
+		config: config,
+	}
+}
+
+// validateGitInput checks if the remote and branch are in the correct format
+func validateGitInput(remote, branch string) error {
+	remoteRE := regexp.MustCompile(gitRemotePattern)
+	if !remoteRE.MatchString(remote) {
+		return fmt.Errorf("%w: invalid git remote format", ErrModOperation)
+	}
+
+	branchRE := regexp.MustCompile(gitBranchPattern)
+	if !branchRE.MatchString(branch) {
+		return fmt.Errorf("%w: invalid git branch format", ErrModOperation)
+	}
+	return nil
+}
+
+// GetGitInfo retrieves the latest commit SHA and timestamp from a Git repository
+func (m *moduleOperator) GetGitInfo(remote, branch string) (string, time.Time, error) {
+	// Validate remote and branch against strict regex patterns before using in exec
+	if err := validateGitInput(remote, branch); err != nil {
+		return "", time.Time{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), gitTimeout)
+	defer cancel()
+
+	// Safe to use remote/branch after validateGitInput ensures they match:
+	// - remote: ^[a-zA-Z0-9][-a-zA-Z0-9_.]*$
+	// - branch: ^[a-zA-Z0-9][-a-zA-Z0-9/_]*$
+	//nolint:gosec // Inputs are validated by regex patterns above
+	cmd := exec.CommandContext(ctx, "git", "ls-remote", remote, "refs/heads/"+branch)
+	out, err := cmd.Output()
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("%w: failed to get SHA: %w", ErrModOperation, err)
+	}
+	if len(out) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: no output from git ls-remote", ErrModOperation)
+	}
+	sha := strings.Split(string(out), "\t")[0]
+	if len(sha) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: empty SHA from git ls-remote", ErrModOperation)
+	}
+
+	cmd = exec.CommandContext(ctx, "git", "show", "-s", "--format=%cI", sha)
+	out, err = cmd.Output()
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("failed to get commit time: %w", err)
+	}
+	if len(out) == 0 {
+		return "", time.Time{}, fmt.Errorf("%w: no output from git show", ErrModOperation)
+	}
+
+	commitTime, err := time.Parse(gitTimeFormat, strings.TrimSpace(string(out)))
+	if err != nil {
+		return "", time.Time{}, fmt.Errorf("failed to parse commit time: %w", err)
+	}
+
+	return sha[:gitSHALength], commitTime, nil
+}
+
+// GetLatestVersion retrieves the latest pseudo-version for a module
+func (m *moduleOperator) GetLatestVersion(modulePath string) (module.Version, error) {
```

Comment:
Where/how is this method used? I don't see any direct reference to this method.

Reply:
Thanks. Updated in a2b1834b52303b7d5483bc639d8e73c8aec98ba4.